File: Neetcode150/java/Leetcode217.java
import java.util.HashMap;
import java.util.HashSet;

public class Leetcode217 {
    public static void main(String[] args) {
        int[] arr1 = {1, 5, 7, 10, 4, 1}; //true
        int[] arr2 = {2, 5, 7, 10, 4, 1}; //false
        int[] arr3 = {3, 7, 7, 10, 0, 9}; //true

        System.out.println(hasDuplicateOn(arr1));
        System.out.println(hasDuplicateOn(arr2));
        System.out.println(hasDuplicateOn(arr3));

        System.out.println(hasDuplicateOn2(arr1));
        System.out.println(hasDuplicateOn2(arr2));
        System.out.println(hasDuplicateOn2(arr3));

    }

    public static boolean hasDuplicateOn(int[] arr) {
        HashSet<Integer> visited = new HashSet<>();

        for (int num : arr) {
            if (visited.contains(num)) {
                return true;
            }
            visited.add(num);
        }
        return false;
    }

    public static boolean hasDuplicateOn2(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = i + 1; j < arr.length; j++) {
                if (arr[i] == arr[j]) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: Neetcode150/golang/leetcode217.go
// CONTAINS DUPLICATE

package main

import "fmt"

func main() {
	arr1 := []int{1, 5, 7, 10, 4, 1}
	arr2 := []int{2, 5, 7, 10, 4, 1}
	arr3 := []int{3, 7, 7, 10, 0, 9}

	fmt.Println(hasDupOn(arr1))
	fmt.Println(hasDupOn(arr2))
	fmt.Println(hasDupOn(arr3))

	fmt.Println(hasDupN2(arr1))
	fmt.Println(hasDupN2(arr2))
	fmt.Println(hasDupN2(arr3))
}

// o(n) solution
func hasDupOn(arr []int) bool {
	visited := make(map[int]bool)

	for _, v := range arr {
		if _, ok := visited[v]; ok {
			return true
		}
		visited[v] = true
	}
	return false
}

// o(n^2) solution
func hasDupN2(arr []int) bool {
	for i := 0; i < len(arr); i++ {
		for i2 := i + 1; i2 < len(arr); i2++ {
			if arr[i] == arr[i2] {
				return true
			}
		}
	}
	return false
}
